src/ctap/mod.rs

```diff
@@ -203,14 +206,19 @@ where
         let mut iv = [0; 16];
         iv.copy_from_slice(&self.rng.gen_uniform_u8x32()[..16]);
 
-        let mut blocks = [[0u8; 16]; 4];
+        let block_len = if cred_random.is_some() { 6 } else { 4 };
+        let mut blocks = vec![[0u8; 16]; block_len];
         blocks[0].copy_from_slice(&sk_bytes[..16]);
         blocks[1].copy_from_slice(&sk_bytes[16..]);
         blocks[2].copy_from_slice(&application[..16]);
         blocks[3].copy_from_slice(&application[16..]);
+        if let Some(cred_random) = cred_random {
+            blocks[4].copy_from_slice(&cred_random[..16]);
+            blocks[5].copy_from_slice(&cred_random[16..]);
+        }
         cbc_encrypt(&aes_enc_key, iv, &mut blocks);
 
-        let mut encrypted_id = Vec::with_capacity(ENCRYPTED_CREDENTIAL_ID_SIZE);
+        let mut encrypted_id = Vec::with_capacity(16 * (block_len + 3));
```

Comment:
Something to keep in mind for a later improvement: maybe we should replace all occurence of `16` constant.
I'm thinking like having something like a `CredentialWrapper` which would provide `encrypt`, `decrypt`, and `block_size` methods for example. This would allow to change the way we wrap them in a more flexible way.

Reply:
Otherwise if we have the constant already defined in some way in the crypto we could import this constant, replace all `16` by this constant and compute the `block_len` above with proper maths.

Reply:
True, this function has never been really nice. I will have to touch it again for 2.1 for a similar modification. I understand moving this change to a later PR is okay?